exchange/bitswap/wantmanager.go

```diff
@@ -0,0 +1,262 @@
+	engine "github.com/ipfs/go-ipfs/exchange/bitswap/decision"
+	bsmsg "github.com/ipfs/go-ipfs/exchange/bitswap/message"
+	bsnet "github.com/ipfs/go-ipfs/exchange/bitswap/network"
+	wantlist "github.com/ipfs/go-ipfs/exchange/bitswap/wantlist"
+	peer "github.com/ipfs/go-ipfs/p2p/peer"
+	u "github.com/ipfs/go-ipfs/util"
+)
+
+type WantManager struct {
+	receiver bsnet.Receiver
+
+	incoming chan []*bsmsg.Entry
+
+	// notification channel for new peers connecting
+	connect chan peer.ID
+
+	// notification channel for peers disconnecting
+	disconnect chan peer.ID
+
+	peers map[peer.ID]*msgQueue
+
+	wl *wantlist.Wantlist
+
+	network bsnet.BitSwapNetwork
+
+	ctx context.Context
+}
+
+func NewWantManager(ctx context.Context, network bsnet.BitSwapNetwork) *WantManager {
+	return &WantManager{
+		incoming:   make(chan []*bsmsg.Entry, 10),
+		connect:    make(chan peer.ID, 10),
+		disconnect: make(chan peer.ID, 10),
+		peers:      make(map[peer.ID]*msgQueue),
+		wl:         wantlist.New(),
+		network:    network,
+		ctx:        ctx,
+	}
+}
+
+type msgPair struct {
+	to  peer.ID
+	msg bsmsg.BitSwapMessage
+}
+
+type cancellation struct {
+	who peer.ID
+	blk u.Key
+}
+
+type msgQueue struct {
+	p peer.ID
+
+	outlk sync.Mutex
+	out   bsmsg.BitSwapMessage
+
+	work chan struct{}
+	done chan struct{}
+}
+
+func (pm *WantManager) WantBlocks(ks []u.Key) {
+	log.Infof("want blocks: %s", ks)
+	pm.addEntries(ks, false)
+}
+
+func (pm *WantManager) CancelWants(ks []u.Key) {
+	pm.addEntries(ks, true)
+}
+
+func (pm *WantManager) addEntries(ks []u.Key, cancel bool) {
+	var entries []*bsmsg.Entry
+	for i, k := range ks {
+		entries = append(entries, &bsmsg.Entry{
+			Cancel: cancel,
+			Entry: wantlist.Entry{
+				Key:      k,
+				Priority: kMaxPriority - i,
+			},
+		})
+	}
+	select {
+	case pm.incoming <- entries:
+	case <-pm.ctx.Done():
+	}
+}
+
+func (pm *WantManager) SendBlock(ctx context.Context, env *engine.Envelope) {
+	// Blocks need to be sent synchronously to maintain proper backpressure
+	// throughout the network stack
+	defer env.Sent()
+
+	msg := bsmsg.New(false)
+	msg.AddBlock(env.Block)
+	log.Infof("Sending block %s to %s", env.Peer, env.Block)
+	err := pm.network.SendMessage(ctx, env.Peer, msg)
+	if err != nil {
+		log.Error(err)
+	}
+}
+
+func (pm *WantManager) startPeerHandler(p peer.ID) *msgQueue {
+	_, ok := pm.peers[p]
+	if ok {
+		// TODO: log an error?
+		return nil
+	}
+
+	mq := newMsgQueue(p)
+
+	// new peer, we will want to give them our full wantlist
+	fullwantlist := bsmsg.New(true)
+	for _, e := range pm.wl.Entries() {
+		fullwantlist.AddEntry(e.Key, e.Priority)
+	}
+	mq.out = fullwantlist
+	mq.work <- struct{}{}
+
+	pm.peers[p] = mq
+	go pm.runQueue(mq)
+	return mq
+}
+
+func (pm *WantManager) stopPeerHandler(p peer.ID) {
+	pq, ok := pm.peers[p]
+	if !ok {
+		// TODO: log error?
+		return
+	}
+
+	close(pq.done)
+	delete(pm.peers, p)
+}
+
+func (pm *WantManager) runQueue(mq *msgQueue) {
```

Comment:
i'd make this a separate function. 
```go
func runQueue(pm *WantManager, mq *msgQueue) { ... }
```
having it on the wantmanager worries me that someone will accidentally touch `pm.peers` directly from that worker or some other sort of race.